src/mono/mono/metadata/bundled-resources.c

```diff
@@ -0,0 +1,193 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+//
+
+#include <glib.h>
+
+#include "bundled-resources-internal.h"
+#include "assembly-internals.h"
+
+#include <mono/metadata/assembly.h>
+#include <mono/metadata/mono-private-unstable.h>
+
+static GHashTable *bundled_resources = NULL;
+
+static int *resources_bundled_counts = NULL;
```

Comment:
We also discussed folding the PDB data directly into the bundled assembly and satellite assembly struct to better model the real dependency. It will still be possible to register a PDB resource without corresponding assembly (not sure that will ever be the case though), just register a bundled assembly resource or a satellite assembly resource with NULL assembly data and 0 len but with PDB data and len set. We should update the symbols search logic to first check if PDB file is in hash table, if not, fallback to old search logic in mono-debug.c. In order to keep the fast optimized, we could have a boolean flags indicating if we added any symbols into the hash map and use that in open_symfile_from_bundle to not search hash map for entries.

Reply:
Updating search logic in assembly.c as well as mono-debug.c to look in hash map will optimize runtime lookup for assemblies, satellite assemblies as well as symbol files. Right now they all do linear search in array/lists based on full string compare of name, but with hash map, lookups will scale much better, so in case where there are larger number of resources bundled, this will improve startup performance as well.

Reply:
Incorporated the new bundle resource retrieval api into the assembly/satellite assembly/symfile bundle search paths, and added bools to early return in `assembly.c` when assemblies or satellite assemblies are added to the hash table.